src/irmin/tree.ml

```diff
@@ -600,9 +602,28 @@ module Make (P : S.PRIVATE) = struct
           match (cached_value x, cached_value y) with
           | Some x, Some y -> equal_node x y
           | _ -> (
-              match (cached_map x, cached_map y) with
-              | Some x, Some y -> map_equal x y
-              | _ -> hash_equal (hash x) (hash y)))
+              match (cached_hash x, cached_hash y) with
+              | Some x, Some y -> equal_hash x y
+              | _ -> (
+                  match (cached_map x, cached_map y) with
+                  | Some x, Some y -> map_equal x y
+                  | _ -> hash_equal (hash x) (hash y))))
+
+    (* same as [equal] but do not compare in-memory maps
+       recursively. *)
+    let maybe_equal (x : t) (y : t) =
```

Comment:
The name is ambiguous, it returns `false` when they still might be equal. 

What about `shallow_equal`?